remappers/hostmetrics/hostmetrics.go

```diff
@@ -0,0 +1,107 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package hostmetrics
+
+import (
+	"path"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+const scopePrefix = "otelcol/hostmetricsreceiver"
+
+var scraperToElasticDataset = map[string]string{
+	"cpu":        "system.cpu",
+	"disk":       "system.diskio",
+	"filesystem": "system.filesystem",
+	"load":       "system.load",
+	"memory":     "system.memory",
+	"network":    "system.network",
+	"paging":     "system.memory",
+	"processes":  "system.process",
+	"process":    "system.process",
+}
+
+// Remapper maps the OTel hostmetrics to Elastic system metrics. These remapped
+// metrics power the curated Kibana dashboards. Each datapoint translated using
+// the remapper is tainted with a boolean attribute `otel_translated`.
+type Remapper struct {
+	logger *zap.Logger
+	cfg    config
+}
+
+// NewRemapper creates a new instance of hostmetrics remapper.
+func NewRemapper(logger *zap.Logger, opts ...Option) *Remapper {
+	return &Remapper{
+		cfg:    newConfig(opts...),
+		logger: logger,
+	}
+}
+
+// Remap remaps an OTel ScopeMetrics to a list of OTel metrics such that the
+// remapped metrics could be trivially converted into Elastic system metrics.
+// The current remapping logic assumes that each Metric in the ScopeMetric
+// will have datapoints for a single timestamp only.
+func (r *Remapper) Remap(src pmetric.ScopeMetrics, out pmetric.MetricSlice) {
+	if !r.Valid(src) {
+		return
+	}
+
+	scope := src.Scope()
+	scraper := path.Base(scope.Name())
+
+	var dataset string // an empty dataset defers setting dataset to the caller
+	if r.cfg.SystemIntegrationDataset {
+		var ok bool
+		dataset, ok = scraperToElasticDataset[scraper]
+		if !ok {
+			r.logger.Warn("no dataset defined for scraper", zap.String("scraper", scraper))
+			return
+		}
+	}
+
+	var err error
+	switch scraper {
+	case "cpu":
+		err = remapCPUMetrics(src.Metrics(), out, dataset)
+	case "memory":
+		err = remapMemoryMetrics(src.Metrics(), out, dataset)
+	case "load":
+		err = remapLoadMetrics(src.Metrics(), out, dataset)
+	case "process":
+		err = remapProcessMetrics(src.Metrics(), out, dataset)
+	default:
+		r.logger.Warn("no remapper found for scope", zap.String("scope", scope.Name()))
```

Comment:
Should we add no-op cases for the scopes that we know don't (yet) have translations? Otherwise there will be a lot of warnings that we can't do anything about...?

Reply:
Good point. I will make this a no-op.